src/main/java/com/googleintern/wfm/ruleengine/action/CsvParser.java

```diff
@@ -0,0 +1,190 @@
+  private static final Pattern ROLESKILL_PATTERN = Pattern.compile("\"skill_id\":\"(\\d+)\"");
+  private static final Pattern PERMISSION_PATTERN =
+      Pattern.compile("\"cases_pool_id\":\"(\\d+)\",\"permission_set_id\":\"(\\d+)\"");
+
+  /**
+   * Read all data using csvFilePath. Parse useful information out and store it in *
+   * userPoolAssignmentList.
+   *
+   * @param csvFilePath : represents the path to the csv file, including the file name.
+   * @return an immutable list of UserPoolAssignmentModel objects
+   * @throws IOException
+   * @throws CsvException
+   */
+  public static ImmutableList<UserPoolAssignmentModel> readFromCSVFile(String csvFilePath)
+      throws IOException, CsvException {
+    // Read all data from input csv file located at given path
+    Reader reader = Files.newBufferedReader(Paths.get(csvFilePath));
+    CSVReader csvReader = new CSVReaderBuilder(reader).withSkipLines(1).build();
+    List<String[]> userRecords = csvReader.readAll();
+
+    // List variable userPoolAssignmentList stores the results reading from the input csv file
+    ImmutableList.Builder<UserPoolAssignmentModel> userPoolAssignmentList =
+        new ImmutableList.Builder<>();
+
+    // Parse data line by line
+    return userRecords.stream()
+        .map(record -> parseData(record))
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  /**
+   * Parse information for a single line of data.
+   *
+   * @param record : String format of one line of data.
+   * @return a UserPoolAssignmentModel object which stores parsing information.
+   */
+  private static UserPoolAssignmentModel parseData(final String[] record) {
+    long userId = Long.parseLong(record[Header.USER_ID.column]);
+    long workforceId = Long.parseLong(record[Header.WORKFORCE_ID.column]);
+    long workgroupId = Long.parseLong(record[Header.WORKGROUP_ID.column]);
+
+    // Parse role ids
+    ImmutableList<Long> roleIds = parseRoleIds(record[Header.ROLE_ID.column]);
+
+    // Parse skill ids
+    ImmutableList<Long> skillIds = parseSkillIds(record[Header.SKILL_ID.column]);
+
+    // Parse role_skills ids
+    ImmutableList<Long> roleSkillIds = parseRoleSkillIds(record[Header.ROLESKILL_ID.column]);
+
+    // Parse pool ids and permission ids
+    Set<PoolAssignmentModel> poolAssignmentsSet =
+        parsePoolAssignments(record[Header.POOL_ASSIGNMENT.column]);
+
+    return UserPoolAssignmentModel.builder()
+        .setUserId(userId)
+        .setWorkforceId(workforceId)
+        .setWorkgroupId(workgroupId)
+        .setRoleIds(roleIds)
+        .setSkillIds(skillIds)
+        .setRoleSkillIds(roleSkillIds)
+        .setPoolAssignments(poolAssignmentsSet)
+        .build();
+  }
+
+  /**
+   * * Parse role ids from an input string variable.
+   *
+   * @param roleIdData : String format of input role ids.
+   * @return an immutable list of long objects that stores role ids.
+   */
+  private static ImmutableList<Long> parseRoleIds(final String roleIdData) {
```

Comment:
Java always makes a copy of primitive parameters before sending them to methods. 
The final prefix on parameter won't really help you in your method.